src/Database/Beam/AutoMigrate/Annotated.hs

```diff
@@ -529,66 +561,110 @@ foreignKeyOnPk ::
+      coerceTheNullableAway :: key (Nullable (Beam.TableField tbl)) -> key (Beam.TableField tbl)
+      coerceTheNullableAway = (unsafeCoerce :: key (Nullable (Beam.TableField tbl)) -> key (Beam.TableField tbl))
+      tName = externalEntity ^. dbEntityDescriptor . dbEntityName
+  in ForeignKey (TableName tName) (S.fromList colPairs)
+
+
+foreignKeyOnNullable ::
+  forall key tbl tbl' be db.
+  ( Beam.Beamable tbl'
+  , Beam.Beamable key
+  ) =>
   DatabaseEntity be db (TableEntity tbl') ->
-  [ForeignKeyConstraint tbl tbl'] ->
-  -- | On Delete
-  ReferenceAction ->
-  -- | On Update
-  ReferenceAction ->
+  -- | fields on local table
+  (tbl (Beam.TableField tbl) -> key (Nullable (Beam.TableField tbl))) ->
+  -- | fields on referenced table
+  (tbl' (Beam.TableField tbl') -> key (Beam.TableField tbl')) ->
   EntityModification (AnnotatedDatabaseEntity be db) be (TableEntity tbl)
-foreignKeyOn externalEntity us onDelete onUpdate =
+foreignKeyOnNullable externalEntity ourColumn theirColumn =
+  foreignKeyOnNullableWithOptions externalEntity ourColumn theirColumn def
+
+foreignKeyOnNullableWithOptions ::
+  forall key tbl tbl' be db.
+  ( Beam.Beamable tbl'
+  , Beam.Beamable key
+  ) =>
+  DatabaseEntity be db (TableEntity tbl') ->
+  -- | fields on local table
+  (tbl (Beam.TableField tbl) -> key (Nullable (Beam.TableField tbl))) ->
+  -- | fields on referenced table
+  (tbl' (Beam.TableField tbl') -> key (Beam.TableField tbl')) ->
+  ForeignKeyConstraintOptions ->
+  EntityModification (AnnotatedDatabaseEntity be db) be (TableEntity tbl)
+foreignKeyOnNullableWithOptions externalEntity ourColumn theirColumn options =
   EntityModification
     ( Endo
         ( \(AnnotatedDatabaseEntity tbl@(AnnotatedDatabaseTable {}) e) ->
             AnnotatedDatabaseEntity
-              ( tbl
-                  { dbAnnotatedConstraints =
-                      let colPairs =
-                            concatMap
-                              ( \case
-                                  (References ours theirs) ->
-                                    zipWith
-                                      (,)
-                                      (fieldAsColumnNames (ours (tableSettings e)))
-                                      [ColumnName (theirs (tableSettings externalEntity) ^. Beam.fieldName)]
-                              )
-                              us
-                          tName = externalEntity ^. dbEntityDescriptor . dbEntityName
-                          conname = T.intercalate "_" (tName : map (columnName . snd) colPairs) <> "_fkey"
-                       in S.insert
-                            (ForeignKey conname (TableName tName) (S.fromList colPairs) onDelete onUpdate)
-                            (dbAnnotatedConstraints tbl)
-                  }
+              ( tbl Lens.& Lens.over (_dbAnnotatedConstraints . _foreignKeyConstraints) (<> M.singleton (mkNullableForeignKeyConstraint e ourColumn externalEntity theirColumn) options)
```

Comment:
I feel like these bits need to go in a lens blog post somewhere :). Crushing it with the line count :D.